File: 백준/Silver/1912. 연속합/연속합.py
import sys

n = int(sys.stdin.readline().rstrip())
arr = list(map(int, sys.stdin.readline().rstrip().split()))

sum_arr = [0] * n
# [값, 인덱스]
answer = arr[0]

sum_arr[0] = arr[0]
min_value = arr[0]
for i in range(1, n):
    sum_arr[i] = sum_arr[i - 1] + arr[i]
    if min_value > sum_arr[i]:
        min_value = sum_arr[i]
    else:
        answer = max(answer, sum_arr[i] - min_value, sum_arr[i], arr[i])
    answer = max(answer, sum_arr[i], arr[i])

# print(sum_arr)
print(answer)

File: 백준/Silver/1912. 연속합/연속합.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int n = Integer.parseInt(br.readLine());
        st = new StringTokenizer(br.readLine(), " ");
        int[] arr = new int[n];
        for(int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        int[] dp = new int[n];
        dp[0] = arr[0];
        int minValue = arr[0];
        int answer = arr[0];
        for(int i = 1; i < n; i++) {
            dp[i] = dp[i - 1] + arr[i];

            if(minValue > dp[i]) {
                minValue = dp[i];
            } else {
                answer = Math.max(answer, Math.max(dp[i] - minValue, Math.max(dp[i], arr[i])));
            }

            answer = Math.max(answer, Math.max(dp[i], arr[i]));
        }

        System.out.println(answer);
    }
}
